aws/acm.go

```diff
@@ -0,0 +1,213 @@
+// example
+func (cc *certificateCache) updateCertCache() {
+	certList, err := cc.ListCerts()
+	if err != nil {
+		log.Printf("Could not update certificate cache, caused by: %v", err)
+		return
+	}
+
+	certDetails := make([]*acm.CertificateDetail, len(certList), len(certList))
+	for idx, o := range certList {
+		certInput := &acm.DescribeCertificateInput{CertificateArn: o.CertificateArn}
+		certDetail, err := cc.acmClient.DescribeCertificate(certInput)
+		if err != nil {
+			log.Printf("Could not get certificate details from AWS for ARN: %s, caused by: %v", o.CertificateArn, err)
+			return
+		}
+		certDetails[idx] = certDetail.Certificate
+	}
+
+	cc.Lock()
+	cc.acmCertSummary = certList
+	cc.acmCertDetail = certDetails
+	cc.Unlock()
+}
+
+// GetCachedCerts returns a copy of the cached acm CertifcateDetail slice
+// https://docs.aws.amazon.com/acm/latest/APIReference/API_ListCertificates.html#API_ListCertificates_RequestSyntax
+// filtered by CertificateStatuses
+// https://docs.aws.amazon.com/acm/latest/APIReference/API_ListCertificates.html#API_ListCertificates_RequestSyntax
+func (cc *certificateCache) GetCachedCerts() []*acm.CertificateDetail {
+	cc.Lock()
+	result := cc.acmCertDetail[:]
+	cc.Unlock()
+	return result
+}
+
+// LsitCerts returns a list of acm Certifcates filtered by
+// CertificateStatuses
+// https://docs.aws.amazon.com/acm/latest/APIReference/API_ListCertificates.html#API_ListCertificates_RequestSyntax
+func (cc *certificateCache) ListCerts() ([]*acm.CertificateSummary, error) {
+	maxItems := aws.Int64(10)
+
+	params := &acm.ListCertificatesInput{
+		CertificateStatuses: []*string{
+			aws.String("ISSUED"), // Required
+		},
+		MaxItems: maxItems,
+	}
+	resp, err := cc.acmClient.ListCertificates(params)
+	if err != nil {
+		return nil, err
+	}
+	certList := resp.CertificateSummaryList
+
+	// more certs if NextToken set in response, use pagination to get more
+	for resp.NextToken != nil {
+		params = &acm.ListCertificatesInput{
+			CertificateStatuses: []*string{
+				aws.String("ISSUED"),
+			},
+			MaxItems:  maxItems,
+			NextToken: resp.NextToken,
+		}
+		resp, err = cc.acmClient.ListCertificates(params)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, cert := range resp.CertificateSummaryList {
+			certList = append(certList, cert)
+		}
+	}
+
+	return certList, nil
+}
+
+// FindBestMatchingCertifcate get all ACM certifcates and use a suffix
+// search best match opertion in order to find the best matching
+// certifcate ARN.
+//
+// We don't need to validate the Revocation here, because we only pull
+// ISSUED certificates.
+func FindBestMatchingCertifcate(certs []*acm.CertificateDetail, hostname string) (*acm.CertificateDetail, error) {
+	candidate := &acm.CertificateDetail{}
+	longestMatch := -1
+	sevenDaysInSec := int64(7 * 24 * 3600)
+
+	for _, cert := range certs {
+		now := time.Now().Unix()
+		notAfter := cert.NotAfter.Unix()
+		notBefore := cert.NotBefore.Unix()
```

Comment:
Shouldn't these 3 statements be outside of the door loop? The issued dates are already set in the cert details. Also, this isn't supposed to take that long.